File: binary_search_tree/BinarySearchTree.java
package binary_search_tree;

import java.util.ArrayDeque;

public class BinarySearchTree<T extends Comparable<T>> {
    private class Node<E extends Comparable<E>> {
        E data;
        Node<E> parent;
        Node<E> left;
        Node<E> right;

        Node(E data) {
            this(data, null);
        }

        Node(E data, Node<E> parent) {
            this(data, parent, null, null);
        }

        Node(E data, Node<E> parent, Node<E> left, Node<E> right) {
            this.data = data;
            this.parent = parent;
            this.left = left;
            this.right = right;
        }
    }

    Node<T> root;
    int size;

    BinarySearchTree() {
        size = 0;
    }

    BinarySearchTree(T data) {
        size = 1;
        root = new Node<T>(data);
    }

    public boolean isEmpty() {
        return root == null;
    }

    public int size() {
        return size;
    }

    public void insert(T data) {
        if (isEmpty()) {
            root = new Node<T>(data);
            return;
        }

        Node<T> previousNode = null;
        Node<T> currentNode = root;

        int comparison = 0;

        while (currentNode != null) {
            previousNode = currentNode;

            comparison = data.compareTo(currentNode.data);

            if (comparison < 0) {
                currentNode = currentNode.left;
            } else if (comparison > 0) {
                currentNode = currentNode.right;
            } else {
                return;
            }
        }

        Node<T> newNode = new Node<>(data, previousNode);

        if (comparison < 0) {
            previousNode.left = newNode;
        } else {
            previousNode.right = newNode;
        }

        size++;
    }

    public boolean contains(T data) {
        Node<T> currentNode = root;

        while (currentNode != null) {
            int comparison = data.compareTo(currentNode.data);

            if (comparison < 0) {
                currentNode = currentNode.left;
            } else if (comparison > 0) {
                currentNode = currentNode.right;
            } else {
                return true;
            }
        }

        return false;
    }

    public void remove(T data) {
        Node<T> currentNode = root;

        while (currentNode != null) {
            int comparison = data.compareTo(currentNode.data);

            if (comparison < 0) {
                currentNode = currentNode.left;
            } else if (comparison > 0) {
                currentNode = currentNode.right;
            } else {
                break;
            }
        }

        if (currentNode == null) {
            return;
        }

        size--;

        if (currentNode.left == null && currentNode.right == null) {
            removeLeaf(currentNode);
        } else if (currentNode.left == null || currentNode.right == null) {
            removeOneChild(currentNode);
        } else {
            removeTwoChild(currentNode);
        }
    }

    private void removeLeaf(Node<T> node) {
        if (node.parent == null) {
            root = null;
        } else if (node == node.parent.left) {
            node.parent.left = null;
        } else {
            node.parent.right = null;
        }
    }

    private void removeOneChild(Node<T> node) {
        if (node == root) {
            if (root.left != null) {
                root = root.left;
            } else {
                root = root.right;
            }
        } else {
            if (node == node.parent.left) {
                if (node.left != null) {
                    node.parent.left = node.left;
                } else {
                    node.parent.left = node.right;
                }
            } else {
                if (node.left != null) {
                    node.parent.right = node.left;
                } else {
                    node.parent.right = node.right;
                }
            }
        }
    }

    private void removeTwoChild(Node<T> node) {
        Node<T> predecessor = node.left;

        while (predecessor.right != null) {
            predecessor = predecessor.right;
        }

        node.data = predecessor.data;

        if (predecessor.left != null || predecessor.right != null) {
            removeOneChild(predecessor);
        } else {
            removeLeaf(predecessor);
        }
    }

    private String inOrderTraverse(Node<T> node) {
        if (node == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        result.append(inOrderTraverse(node.left)).append(node.data).append(" ")
                .append(inOrderTraverse(node.right));

        return result.toString();
    }

    public String inOrder() {
        return inOrderTraverse(root);
    }

    public String toString() {
        return inOrder();
    }

    private String preOrderTraverse(Node<T> node) {
        if (node == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        result.append(node.data).append(" ").append(inOrderTraverse(node.left))
                .append(inOrderTraverse(node.right));

        return result.toString();
    }

    public String preOrder() {
        return preOrderTraverse(root);
    }

    private String postOrderTraverse(Node<T> node) {
        if (node == null) {
            return "";
        }

        StringBuilder result = new StringBuilder();

        result.append(inOrderTraverse(node.left))
                .append(inOrderTraverse(node.right)).append(node.data).append(" ");

        return result.toString();
    }

    public String postOrder() {
        return postOrderTraverse(root);
    }

    public String levelOrder() {
        if (isEmpty()) {
            return "";
        }

        ArrayDeque<Node<T>> queue = new ArrayDeque<>();
        StringBuilder result = new StringBuilder();

        queue.add(root);

        while (!queue.isEmpty()) {
            Node<T> node = queue.pollFirst();

            result.append(node.data).append(" ");

            if (node.left != null) {
                queue.add(node.left);
            }

            if (node.right != null) {
                queue.add(node.right);
            }
        }

        return result.toString();
    }

    /*
     * The following method is unnecessary because of Java's garbage collector,
     * but is included because it'll be a helpful guide when implementing the
     * C version:
     * 
     * private void postOrderRemove(Node<T> node) {
     * if (node == null) {
     * return;
     * }
     * 
     * postOrderRemove(node.left);
     * postOrderRemove(node.right);
     * 
     * if (node == root) {
     * root = null;
     * } else if (node == node.parent.left) {
     * node.parent.left = null;
     * } else {
     * node.parent.right = null;
     * }
     * }
     */

    public void clear() {
        // postOrderRemove(root); // Unnecessary because of Java's garbage collector,
        // but the line is included for completion's sake (and because it'll be
        // necessary in the C implementation).
        root = null;
        size = 0;
    }

    public static void main(String[] args) {
        BinarySearchTree<Integer> myBST = new BinarySearchTree<>();

        for (int i = 0; i < 15; ++i) {
            myBST.insert((17 * (i + 1) + 3) % 31);
        }

        System.out.println("Pre-order traversal: " + myBST.preOrder());
        System.out.println("In-order traversal: " + myBST);
        System.out.println("Post-order traversal: " + myBST.postOrder());
        System.out.println("Level-order traversal: " + myBST.levelOrder());
        System.out.println("Size: " + myBST.size());
        System.out.println("Empty? " + myBST.isEmpty());

        System.out.println("\nContains 7? " + myBST.contains(7));
        System.out.println("\nContains 11? " + myBST.contains(11));

        System.out.println("\nTesting removal of leaf (18).");
        myBST.remove(18);
        System.out.println("Level-order traversal: " + myBST.levelOrder());

        System.out.println("\nTesting removal of node with one child (1).");
        myBST.remove(1);
        System.out.println("Level-order traversal: " + myBST.levelOrder());

        System.out.println("\nTesting removal of node with two children (20).");
        myBST.remove(20);
        System.out.println("Level-order traversal: " + myBST.levelOrder());

        System.out.println("\nClearing tree. The traversals should be blank below this line.");
        myBST.clear();

        System.out.println("Pre-order traversal: " + myBST.preOrder());
        System.out.println("In-order traversal: " + myBST);
        System.out.println("Post-order traversal: " + myBST.postOrder());
        System.out.println("Level-order traversal: " + myBST.levelOrder());
        System.out.println("Size: " + myBST.size());
        System.out.println("Empty? " + myBST.isEmpty());
    }
}

File: binary_search_tree/binary_search_tree.c
#include "binary_search_tree.h"

bst_t *create_bst(int (*comparator)(void *, void *),
                  void (*printer)(void *),
                  void (*deleter)(void *))
{
    bst_t *tree = malloc(sizeof(bst_t));

    if( !tree )
    {
        return nullptr;
    }

    tree->root = nullptr;
    tree->size = 0;
    tree->comparator = comparator;
    tree->printer = printer;
    tree->deleter = deleter;

    return tree;
}

node_t *create_node(void *data, node_t *parent)
{
    node_t *node = malloc(sizeof(node_t));

    if( !node )
    {
        return nullptr;
    }

    node->data = data;
    node->left = node->right = nullptr;
    node->parent = parent;

    return node;
}

node_t *find_previous_node(bst_t *tree, void *data)
{
    if( !tree )
    {
        return nullptr;
    }

    node_t *previous_node = nullptr;
    node_t *current_node = tree->root;

    int comparison;

    while( current_node )
    {
        previous_node = current_node;

        comparison = tree->comparator(data, current_node->data);

        if( comparison < 0 )
        {
            current_node = current_node->left;
        }
        else if( comparison > 0 )
        {
            current_node = current_node->right;
        }
        else
        {
            return nullptr;
        }
    }

    return previous_node;
}

node_t *find_node(bst_t *tree, void *data)
{
    if( !tree )
    {
        return nullptr;
    }

    node_t *current_node = tree->root;

    while( current_node )
    {
        int comparison = tree->comparator(data, current_node->data);

        if( comparison < 0 )
        {
            current_node = current_node->left;
        }
        else if( comparison > 0 )
        {
            current_node = current_node->right;
        }
        else
        {
            return current_node;
        }
    }

    return nullptr;
}

void insert_node(bst_t *tree, void *data)
{
    if( is_empty(tree) )
    {
        tree->root = create_node(data, nullptr);
        ++tree->size;
        return;
    }

    node_t *previous_node = find_previous_node(tree, data);

    node_t *node = create_node(data, previous_node);

    if( tree->comparator(data, previous_node->data) < 0 )
    {
        previous_node->left = node;
    }
    else
    {
        previous_node->right = node;
    }

    ++tree->size;
}

bool contains(bst_t *tree, void *data)
{
    return find_node(tree, data);
}

void *remove_node(bst_t *tree, void *data)
{
    if( is_empty(tree) )
    {
        return nullptr;
    }

    node_t *current_node = find_node(tree, data);

    if( !current_node )
    {
        return nullptr;
    }

    --tree->size;

    if( current_node->left && current_node->right )
    {
        return remove_two_child(tree, current_node);
    }
    else if( current_node->left || current_node->right )
    {
        return remove_one_child(tree, current_node);
    }

    return remove_leaf(tree, current_node);
}

void *remove_leaf(bst_t *tree, node_t *node)
{
    if( is_empty(tree) || !node )
    {
        return nullptr;
    }

    if( node == tree->root )
    {
        tree->root = nullptr;
    }
    else if( node->parent->left == node )
    {
        node->parent->left = nullptr;
    }
    else
    {
        node->parent->right = nullptr;
    }

    void *data = node->data;
    free(node);
    return data;
}

void *remove_one_child(bst_t *tree, node_t *node)
{
    if( is_empty(tree) || !node || (node->left && node->right) )
    {
        return nullptr;
    }

    node_t **relink_node;
    if( node == tree->root )
    {
        relink_node = &tree->root;
    }
    else
    {
        relink_node = (node == node->parent->left) ? &node->parent->left
                                                   : &node->parent->right;
    }

    *relink_node = (node->left) ? node->left : node->right;

    void *data = node->data;
    free(node);
    return data;
}

void *remove_two_child(bst_t *tree, node_t *node)
{
    if( is_empty(tree) || !node )
    {
        return nullptr;
    }

    node_t *predecessor = node->left;

    while( predecessor->right )
    {
        predecessor = predecessor->right;
    }

    void *data = node->data;
    node->data = predecessor->data;

    if( !predecessor->left && !predecessor->right )
    {
        remove_leaf(tree, predecessor);
    }
    else
    {
        remove_one_child(tree, predecessor);
    }

    return data;
}

void clear(bst_t *tree)
{
    if( !tree )
    {
        return;
    }

    remove_recursive(tree->root, tree->deleter);
    tree->root = nullptr;
    tree->size = 0;
}

void remove_recursive(node_t *node, void (*deleter)(void *))
{
    if( !node )
    {
        return;
    }

    remove_recursive(node->left, deleter);
    remove_recursive(node->right, deleter);

    deleter(node->data);
    free(node);
}

void pre_order_print(bst_t *tree)
{
    if( !tree )
    {
        return;
    }

    pre_order_recurse(tree->root, tree->printer);
    printf("\n\n");
}

void pre_order_traverse(bst_t *tree, void (*func)(void *))
{
    if( !tree )
    {
        return;
    }

    pre_order_recurse(tree->root, func);
}

void pre_order_recurse(node_t *node, void (*func)(void *))
{
    if( !node )
    {
        return;
    }

    func(node->data);
    pre_order_recurse(node->left, func);
    pre_order_recurse(node->right, func);
}

void in_order_print(bst_t *tree)
{
    if( !tree )
    {
        return;
    }

    in_order_recurse(tree->root, tree->printer);
    printf("\n\n");
}

void in_order_traverse(bst_t *tree, void (*func)(void *))
{
    if( !tree )
    {
        return;
    }

    in_order_recurse(tree->root, func);
}

void in_order_recurse(node_t *node, void (*func)(void *))
{
    if( !node )
    {
        return;
    }

    in_order_recurse(node->left, func);
    func(node->data);
    in_order_recurse(node->right, func);
}

void post_order_print(bst_t *tree)
{
    if( !tree )
    {
        return;
    }

    post_order_recurse(tree->root, tree->printer);
    printf("\n\n");
}

void post_order_traverse(bst_t *tree, void (*func)(void *))
{
    if( !tree )
    {
        return;
    }

    post_order_recurse(tree->root, func);
}

void post_order_recurse(node_t *node, void (*func)(void *))
{
    if( !node )
    {
        return;
    }

    post_order_recurse(node->left, func);
    post_order_recurse(node->right, func);
    func(node->data);
}

void level_order_print(bst_t *tree)
{
    if( !tree )
    {
        return;
    }

    level_order_traverse(tree, tree->printer);
    printf("\n\n");
}

void level_order_traverse(bst_t *tree, void (*func)(void *))
{
    if( !tree || !tree->root )
    {
        return;
    }

    deque_t *queue = CREATE_DEQUE(node_t *);

    push_back(queue, &tree->root);

    while( queue->size > 0 )
    {
        node_t *node;
        pop_front(queue, &node);

        if( !node )
        {
            continue;
        }

        func(node->data);

        if( node->left )
        {
            push_back(queue, &(node->left));
        }

        if( node->right )
        {
            push_back(queue, &node->right);
        }
    }

    delete_deque(&queue);
}

bool is_empty(bst_t *tree)
{
    return !tree || tree->size == 0;
}

void delete_bst(bst_t **tree)
{
    if( !tree || !*tree )
    {
        return;
    }

    clear(*tree);
    free(*tree);
    *tree = nullptr;
}
